File: frontend/src/stores/reviews.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '@/api/client'
import type { Review, ReviewUpdate } from '@/types'

export const useReviewsStore = defineStore('reviews', () => {
  const reviews = ref<Review[]>([])
  const saving = ref(false)

  function getReview(insightId: string): Review | undefined {
    return reviews.value.find(r => r.insight_id === insightId)
  }

  async function fetchReviews() {
    reviews.value = await api.getReviews()
  }

  async function saveReview(insightId: string, update: ReviewUpdate) {
    saving.value = true
    try {
      const saved = await api.updateReview(insightId, update)
      const idx = reviews.value.findIndex(r => r.insight_id === insightId)
      if (idx >= 0) reviews.value[idx] = saved
      else reviews.value.push(saved)
    } finally {
      saving.value = false
    }
  }

  return { reviews, saving, getReview, fetchReviews, saveReview }
})

File: api/routers/reviews.py
from fastapi import APIRouter

import api.data as data_module
from api.models import Review, ReviewUpdate

router = APIRouter()


@router.get("/api/reviews", response_model=list[Review])
def list_reviews() -> list[Review]:
    return data_module.load_review_log(data_module.DEFAULT_DATA_DIR)


@router.patch("/api/reviews/{insight_id}", response_model=Review)
def update_review(insight_id: str, update: ReviewUpdate) -> Review:
    return data_module.save_review(insight_id, update, data_module.DEFAULT_DATA_DIR)
